.circleci/config.yml

```diff
@@ -0,0 +1,119 @@
+defaults: &defaults
+    docker:
+      - image: circleci/openjdk:8-jdk-stretch
+    environment:
+      MAVEN_OPTS: -Xmx3200m
+      _JAVA_OPTIONS: "-Xms1g -Xmx1g"
+
+version: 2
+jobs:
+  closure_library:
+    <<: *defaults
+    working_directory: ~/
+    steps:
+      - restore_cache:
+          keys:
+            - v1-closure-library
+      - run:
+          name: 
+          command: |
+            if [ -e closure-library ]; then
+              # if the cache contained it, let's just update
+              echo update from remote
+
+              cd closure-library
+              git remote update
+              git pull
+            else
+              echo cloning repository
+              git clone git@github.com:google/closure-library.git
+            fi
+      - save_cache:
+          key: v1-closure-library
```

Comment:
Don't do that.

From https://circleci.com/docs/2.0/caching/#writing-to-the-cache-in-workflows

> Cache is immutable on write: once a cache is written for a particular key like node-cache-master, it cannot be written to again.

and https://circleci.com/docs/2.0/configuration-reference/#save_cache

> The cache for a specific key is immutable and cannot be changed once written.
> 
> *Note* If the cache for the given key already exists it won’t be modified, and job execution will proceed to the next step.

You'd probably want to use something like `git ls-tree HEAD > ../closure-library.tree` and then `key: v1-closure-library-{checksum "closure-library.tree"}` (you'll want to keep the `restore_cache` with a prefix, so you get the most recent cache matching `v1-closure-library`, then `git pull` it and `save_cache` it under a new name, if it changed; taking advantage of the write-only nature of the cache)

Reply:
Thanks!   It's been a long while since I configured circleci and forgot about that.  That would have gone unnoticed for awhile, so thanks for the catch.

Reply:
The problem is chicken/egg.  We cannot use the ls-tree of the closure-library checkout if we don't currently have the checkout to do that on, so we have to base the key upon something within the build.    Since cirlceci restore_cache uses fallback, I'll at least be able to supply multiple keys and a fallback to receive the "latest" cache.

Then, when saving, I could just use something like the CIRCLE_SHA variable - which will at least make sure that the other jobs within the same build will get the same cache.

So, I was thinking something along the lines of the following:
```yaml
      - run:
          name: create cache key 
          command: echo $CIRCLE_SHA1 > .circle-sha
      - restore_cache:
          keys:
            - v1-closure-library-{{ checksum ".circle-sha" }}
            # if cache is not present load any most recent
            - v1-closure-library-
```

Of course, on initial run in the build, it WON'T have a cache for the current $CIRCLE_SHA1, so it'll always have to fallback to the most recent.  It'll do a git pull and save the cache specific to the current build.